url400631cdd0_ludaiqian_appliedxml_tgz-pJ8-test_simples_TestJ8/src/com/google/gsoncode/internal/ConstructorConstructor.java

```diff
@@ -0,0 +1,209 @@
+import java.lang.reflect.InvocationTargetException;
+import java.lang.reflect.ParameterizedType;
+import java.lang.reflect.Type;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.LinkedHashMap;
+import java.util.LinkedHashSet;
+import java.util.LinkedList;
+import java.util.Map;
+import java.util.Queue;
+import java.util.Set;
+import java.util.SortedMap;
+import java.util.SortedSet;
+import java.util.TreeMap;
+import java.util.TreeSet;
+
+
+/**
+ * Returns a function that can construct an instance of a requested type.
+ */
+public final class ConstructorConstructor {
+	private final Map<Type, InstanceCreator<?>> instanceCreators;
+
+	public ConstructorConstructor() {
+		instanceCreators=Collections.<Type, InstanceCreator<?>>emptyMap();
+	}
+
+	public ConstructorConstructor(Map<Type, InstanceCreator<?>> instanceCreators) {
+		this.instanceCreators = instanceCreators;
+	}
+
+	public <T> ObjectConstructor<T> get(TypeToken<T> typeToken) {
+		final Type type = typeToken.getType();
+		final Class<? super T> rawType = typeToken.getRawType();
+
+		// first try an instance creator
+
+		@SuppressWarnings("unchecked")
+		// types must agree
+		final InstanceCreator<T> typeCreator = (InstanceCreator<T>) instanceCreators.get(type);
+		if (typeCreator != null) {
+			return new ObjectConstructor<T>() {
+				public T construct() {
+					return typeCreator.createInstance(type);
+				}
+			};
+		}
+
+		// Next try raw type match for instance creators
+		@SuppressWarnings("unchecked")
+		// types must agree
+		final InstanceCreator<T> rawTypeCreator = (InstanceCreator<T>) instanceCreators.get(rawType);
+		if (rawTypeCreator != null) {
+			return new ObjectConstructor<T>() {
+				public T construct() {
+					return rawTypeCreator.createInstance(type);
+				}
+			};
+		}
+
+		ObjectConstructor<T> defaultConstructor = newDefaultConstructor(rawType);
+		if (defaultConstructor != null) {
+			return defaultConstructor;
+		}
+
+		ObjectConstructor<T> defaultImplementation = newDefaultImplementationConstructor(type, rawType);
+		if (defaultImplementation != null) {
+			return defaultImplementation;
+		}
+
+		// finally try unsafe
+		return newUnsafeAllocator(type, rawType);
+	}
+
+	private <T> ObjectConstructor<T> newDefaultConstructor(Class<? super T> rawType) {
+		try {
+			final Constructor<? super T> constructor = rawType.getDeclaredConstructor();
+			if (!constructor.isAccessible()) {
+				constructor.setAccessible(true);
+			}
+			return new ObjectConstructor<T>() {
+				@SuppressWarnings("unchecked")
+				// T is the same raw type as is requested
+				public T construct() {
+					try {
+						Object[] args = null;
+						return (T) constructor.newInstance(args);
+					} catch (InstantiationException e) {
+						// TODO: JsonParseException ?
+						throw new RuntimeException("Failed to invoke " + constructor + " with no args", e);
+					} catch (InvocationTargetException e) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: While wrapping the caught exception into a custom one, information about the caught exception is being lost, including information about the stack trace of the exception.

**Fix**: If the caught exception object does not contain sensitive information, consider passing it as the &#34;rootCause&#34; or inner exception parameter to the constructor of the new exception before throwing the new exception. (Note that not all exception constructors support inner exceptions. Use a wrapper exception that supports inner exceptions.) 
[Learn more](https://www.ibm.com/support/pages/best-practice-catching-and-re-throwing-java-exceptions)